synapse/util/logcontext.py

```diff
@@ -346,13 +345,38 @@ def get_resource_usage(self):
         # can include resource usage so far.
         is_main_thread = threading.current_thread() is self.main_thread
         if self.alive and self.usage_start and is_main_thread:
-            current = get_thread_resource_usage()
-            res.ru_utime += current.ru_utime - self.usage_start.ru_utime
-            res.ru_stime += current.ru_stime - self.usage_start.ru_stime
+            utime_delta, stime_delta = self._get_cputime()
+            res.ru_utime += utime_delta
+            res.ru_stime += stime_delta
 
         return res
 
+    def _get_cputime(self):
+        """Get the cpu usage time so far
+
+        Returns: Tuple[float, float]: seconds in user mode, seconds in system mode
+        """
+        current = get_thread_resource_usage()
+
+        utime_delta = current.ru_utime - self.usage_start.ru_utime
+        stime_delta = current.ru_stime - self.usage_start.ru_stime
+
+        # sanity check
+        if utime_delta < 0:
+            raise ValueError("utime went backwards! %f < %f" % (
+                current.ru_utime, self.usage_start.ru_utime,
+            ))
```

Comment:
What's the effect of raising here? Will this just take out metrics? Do we want to log.error but return 0 instead?

Reply:
No, it will take out whatever is running at this point.

My concern with `log.error` is that it's never going to get noticed, and if it is, we won't be able to figure out why it is happening.